### PR TITLE
[ros2doctor] Add QoS compatibility check and report

### PR DESCRIPTION
**Pull request by audrow**
Here is an example of what the check looks like if an incompatibility error is detected;
```
/root/ros2_ws/build/ros2doctor/ros2doctor/api/qos_compatibility.py: 49: UserWarning: ERROR: QOS compatibility error found on topic '/my_chatter'. Use `ros2 doctor --report` for more details.
```

Here is what a report looks like with two pub/sub pairs, one compatible and one incompatible:
```
   QOS COMPATIBILITY LIST
topic [type]            : /chatter [std_msgs/msg/String]
publisher node          : talker
subscriber node         : listener
compatibility status    : OK
topic [type]            : /my_chatter [std_msgs/msg/String]
publisher node          : mynode
subscriber node         : mynode
compatibility status    : ERROR: Best effort publisher and reliable subscription;
```

The messages for warnings are similar.

Here is the report if there are no publisher/subscription pairs:
```
   QOS COMPATIBILITY LIST
compatibility status    : No publisher/subscriber pairs found
```

I haven't written any tests for this yet and would be happy for suggestions on how to do so.

### CODE REVIEW

**Review by fujitatomoya — COMMENTED** [review]
i am good to go with this, test would be ideal.

**Review by jacobperron — COMMENTED** [review]
Looks good! I've left some comments/suggestions below.

I haven't tried it out yet. I agree that it would be good to add a test.

File: ros2doctor/ros2doctor/api/__init__.py [hunk 1]
```diff
@@ -62,6 +63,10 @@ def add_to_report(self, item_name: str, item_info: str) -> None:
         """Add report content to items list (list of string tuples)."""
         self.items.append((item_name, item_info))
 
+    def add_section_to_report(self, section_name: str) -> None:
+        """Add report content to items list (list of string tuples)."""
```
[the remark this answers is not in the document]
Comment (jacobperron) [reply ros2doctor/ros2doctor/api/__init__.py:67]: It doesn't look like this function is called anywhere. Maybe it can be removed.

**Comment by audrow** [comment]
The new check error output is as follows:
```
/root/ros2_ws/build/ros2doctor/ros2doctor/api/qos_compatibility.py: 49: UserWarning: ERROR: QoS compatibility error found on topic '/chatter': Best effort publisher and reliable subscription;
```

Warnings are displayed similarly.

**Review by jacobperron — APPROVED** [review]
LGTM pending CI

File: ros2doctor/test/fixtures/listener_node_with_reliable_qos.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
+# Copyright 2019 Open Source Robotics Foundation, Inc.
```
Comment (fujitatomoya) [inline ros2doctor/test/fixtures/listener_node_with_reliable_qos.py:1]:
```suggestion
# Copyright 2021 Open Source Robotics Foundation, Inc.
```

File: ros2doctor/test/fixtures/talker_node_with_best_effort_qos.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
+# Copyright 2019 Open Source Robotics Foundation, Inc.
```
Comment (fujitatomoya) [inline ros2doctor/test/fixtures/talker_node_with_best_effort_qos.py:1]:
```suggestion
# Copyright 2021 Open Source Robotics Foundation, Inc.
```

File: ros2doctor/test/fixtures/talker_node_with_reliable_qos.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
+# Copyright 2019 Open Source Robotics Foundation, Inc.
```
Comment (fujitatomoya) [inline ros2doctor/test/fixtures/talker_node_with_reliable_qos.py:1]:
```suggestion
# Copyright 2021 Open Source Robotics Foundation, Inc.
```

**Review by fujitatomoya — APPROVED** [review]
lgtm with a couple of nitpicks.

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14380/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9155/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12056/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14479/)

File: ros2doctor/test/fixtures/listener_node_with_reliable_qos.py (hunk 2 above)
Comment (audrow) [reply ros2doctor/test/fixtures/listener_node_with_reliable_qos.py]: Good catch, thanks!

**Comment by audrow** [comment]
Another CI run:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14385/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9160/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12061/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14484/)

**Comment by audrow** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14389/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9163/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12064/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14487/)

**Comment by audrow** [comment]
I weakened the tests because I was still getting the errors on CI with repeated retries. I'm not sure why several ROS2 doctor checks fail on CI, but not on my local machine.  

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14400/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9173/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12074/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14498/)

**Review by jacobperron — APPROVED** [review]

**Comment by RDaneelOlivav** [comment]
Is this feture in the debians already? Because I've run this in my local isntalation og Galactic and it just doesnt generate this Compatibility info. Any ideas?

**Comment by fujitatomoya** [comment]
I think that this is merged only to master (Humble), so galactic does not support this feature.

**Comment by jacobperron** [comment]
That's correct, this is only available in Rolling (to be Humble). However, I think the nature of the change makes it backportable if someone would like to open up a backport PR to Galactic (and maybe Foxy).

**Pull request merged**